File: src/ollie/event.py
"""Stable SDK event envelope (Layer 1)."""

from __future__ import annotations

from typing import Any
from uuid import uuid4

from ollie.trace import utc_now_iso

EVENT_TYPE_TRACE_VALIDATE = "sdk.trace.validate"
EVENT_TYPE_TRACE_PROCESS = "sdk.trace.process"
EVENT_TYPE_TRACE_INGEST = "sdk.trace.ingest"


def new_event_id() -> str:
    return str(uuid4())


def session_id_from_payload(payload: dict[str, Any]) -> str:
    cid = str(payload.get("conversation_id") or "").strip()
    if cid:
        return cid
    aid = str(payload.get("agent_id") or "").strip()
    return aid or "session-unknown"


def build_event(
    *,
    event_type: str,
    payload: dict[str, Any],
    event_id: str | None = None,
) -> dict[str, Any]:
    agent_id = str(payload.get("agent_id") or "").strip()
    if not agent_id:
        raise ValueError("payload.agent_id is required")
    return {
        "event_id": event_id or new_event_id(),
        "agent_id": agent_id,
        "session_id": session_id_from_payload(payload),
        "timestamp": utc_now_iso(),
        "event_type": event_type,
        "payload": payload,
    }

File: packages/ts/src/event.ts
import { randomUUID } from "node:crypto";
import { utcNowIso } from "./version.js";

export const EVENT_TYPE_TRACE_VALIDATE = "sdk.trace.validate";
export const EVENT_TYPE_TRACE_PROCESS = "sdk.trace.process";
export const EVENT_TYPE_TRACE_INGEST = "sdk.trace.ingest";

export function newEventId(): string {
  return randomUUID();
}

export function sessionIdFromPayload(payload: Record<string, unknown>): string {
  const sid = String(payload.session_id ?? "").trim();
  if (sid) return sid;
  const cid = String(payload.conversation_id ?? "").trim();
  if (cid) return cid;
  const aid = String(payload.agent_id ?? "").trim();
  return aid || "session-unknown";
}

export function buildEvent(options: {
  eventType: string;
  payload: Record<string, unknown>;
  eventId?: string;
}): Record<string, unknown> {
  const agentId = String(options.payload.agent_id ?? "").trim();
  if (!agentId) {
    throw new Error("payload.agent_id is required");
  }
  return {
    event_id: options.eventId ?? newEventId(),
    agent_id: agentId,
    session_id: sessionIdFromPayload(options.payload),
    timestamp: utcNowIso(),
    event_type: options.eventType,
    payload: options.payload,
  };
}
